### PR TITLE
MultiAuth

### LINKED ISSUE #209: Enable setting default account for Instances
Would be a nice feature aswell.

### PR DESCRIPTION
**Pull request by 02JanDal**
This PR adds support for multiple account types. Most of the code originates from MultiLaunch.

It also includes a system for resources (currently used for icons).

Things that have been tested:
* Adding/removing minecraft account
* Launching instances
* Setting default account globally and per-instance (closes #209)

Since the format of the accounts.json file is changed a one-way migration is performed and the old one backed up.

### CODE REVIEW

File: application/AuthTask.cpp [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+			// easiest: have default and it's valid
+			m_account = account;
+			emitSucceeded();
+			return;
+		}
+		catch (...)
+		{
+			setStatus(tr("Querying user for authentication details..."));
+			emit progress(101, 100);
+			AccountLoginDialog dlg(account);
```
Comment (peterix) [inline application/AuthTask.cpp]: Tasks doing UI : no.

File: .travis.yml [hunk 2]
```diff
@@ -4,18 +4,18 @@ compiler:
...
 cache: apt
 before_install:
- - sudo apt-add-repository -y ppa:beineri/opt-qt521
+ - sudo apt-add-repository -y ppa:beineri/opt-qt541
  - sudo apt-add-repository -y ppa:kalakris/cmake
  - sudo add-apt-repository -y ppa:ubuntu-toolchain-r/test
  - sudo apt-get update -qq
 install:
- - sudo apt-get install -y -qq cmake qt52base qt52svg qt52tools qt52x11extras qt52webkit
+ - sudo apt-get install -y -qq cmake qt54base qt54svg qt54tools qt54x11extras qt54webkit
```
Comment (peterix) [inline .travis.yml]: How did this get mixed in here?

File: application/AuthTask.cpp [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+		try
+		{
+			setStatus(tr("Checking if default account is valid..."));
+			runTask(account->createCheckTask(m_session));
+			// easiest: have default and it's valid
+			m_account = account;
+			emitSucceeded();
+			return;
+		}
+		catch (...)
```
Comment (peterix) [inline application/AuthTask.cpp]: Tasks using exceptions: NO.

File: application/AuthTask.cpp [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+void AuthTask::executeTask()
+{
+	BaseAccount *account = MMC->accountsModel()->get(m_type, m_instance);
+	if (account)
+	{
+		try
+		{
+			setStatus(tr("Checking if default account is valid..."));
+			runTask(account->createCheckTask(m_session));
```
Comment (peterix) [inline application/AuthTask.cpp]: This is not a good way of running a task. It provides absolutely no feedback.

File: application/CMakeLists.txt [hunk 5]
```diff
@@ -333,7 +335,8 @@ if(WIN32)
 	set(MultiMC_LINK_ADDITIONAL_LIBS ${MultiMC_LINK_ADDITIONAL_LIBS} Qt5::WinMain)
 endif(WIN32)
 
-include_directories(../logic)
+include_directories(${CMAKE_SOURCE_DIR}/logic ${CMAKE_CURRENT_SOURCE_DIR})
```
Comment (peterix) [inline application/CMakeLists.txt]: What?

File: application/MainWindow.cpp [hunk 6]
```diff
@@ -1563,95 +1491,40 @@ void MainWindow::on_actionLaunchInstanceOffline_triggered()
...
 	if (!m_selectedInstance)
+	{
 		return;
+	}
 
-	// Find an account to use.
-	std::shared_ptr<MojangAccountList> accounts = MMC->accounts();
-	MojangAccountPtr account = accounts->activeAccount();
-	if (accounts->count() <= 0)
+	while (true)
```
Comment (peterix) [inline application/MainWindow.cpp]: This is FUNDAMENTALLY BROKEN compared to the original. There is no way to cancel out of this endless loop, only by actually providing a password (WHEN the auth service works, it doesn't have to). REVERT.

File: application/MainWindow.cpp [hunk 7]
```diff
@@ -1563,95 +1491,40 @@ void MainWindow::on_actionLaunchInstanceOffline_triggered()
...
-			EditAccountDialog passDialog(failReason, this, EditAccountDialog::PasswordField);
-			if (passDialog.exec() == QDialog::Accepted)
-			{
-				password = passDialog.password();
-			}
-			else
-			{
-				tryagain = false;
-			}
+			qCritical() << "This should not be reachable";
```
Comment (peterix) [inline application/MainWindow.cpp]: This is VERY reachable.

File: application/MainWindow.cpp [hunk 8]
```diff
@@ -1856,11 +1705,6 @@ void MainWindow::selectionBad()
 	setSelectedInstanceById(MMC->settings()->get("SelectedInstance").toString());
 }
 
-void MainWindow::instanceEnded()
-{
-	this->show();
```
Comment (peterix) [inline application/MainWindow.cpp]: WHAT!?

File: application/dialogs/AccountLoginDialog.cpp [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+void AccountLoginDialog::setSession(SessionPtr session)
+{
+	m_session = session;
+}
+
+void AccountLoginDialog::on_cancelBtn_clicked()
+{
+	reject();
```
Comment (peterix) [inline application/dialogs/AccountLoginDialog.cpp]: NOTE: reject is a REJECT. By the USER. You don't go asking again in an infinite loop.

File: application/dialogs/AccountSelectDialog.cpp [hunk 10]
```diff
@@ -1,85 +0,0 @@
-/* Copyright 2013-2015 MultiMC Contributors
- *
```
Comment (peterix) [inline application/dialogs/AccountSelectDialog.cpp]: What was so wrong with this? Also, it's missing.

File: application/pages/VersionPage.cpp [hunk 11]
```diff
@@ -40,9 +40,9 @@
 #include "tasks/Task.h"
 
 #include "minecraft/MinecraftProfile.h"
-#include "auth/MojangAccountList.h"
 #include "minecraft/Mod.h"
 #include "icons/IconList.h"
+#include "Exception.h"
```
Comment (peterix) [inline application/pages/VersionPage.cpp]: NO?

File: application/widgets/AccountsWidget.ui [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+    <x>0</x>
+    <y>0</y>
+    <width>640</width>
+    <height>480</height>
+   </rect>
+  </property>
+  <property name="windowTitle">
+   <string>Accounts</string>
+  </property>
+  <layout class="QGridLayout" name="gridLayout">
```
Comment (peterix) [inline application/widgets/AccountsWidget.ui]: This has some ugly margins. They should be 0.

File: application/widgets/AccountsWidget.ui [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+   <item row="0" column="0">
+    <widget class="QTreeView" name="view"/>
```
Comment (peterix) [inline application/widgets/AccountsWidget.ui]: Right column should resize to content and be aligned to the right side, left column should expand automatically.

File: logic/AbstractCommonModel.cpp [hunk 14]
```diff
@@ -0,0 +1,102 @@
+// Licensed under the Apache-2.0 license. See README.md for details.
+
+#include "AbstractCommonModel.h"
+
```
Comment (peterix) [inline logic/AbstractCommonModel.cpp]: I really don't get what this is for.

File: logic/BaseInstance.h [hunk 15]
```diff
@@ -23,7 +23,7 @@
 
 #include "settings/INIFile.h"
 #include "BaseVersionList.h"
-#include "auth/MojangAccount.h"
+#include "minecraft/auth/MojangAccount.h"
```
Comment (peterix) [inline logic/BaseInstance.h]: BaseInstance.... Mojang... ????

File: logic/CMakeLists.txt [hunk 16]
```diff
@@ -52,6 +61,20 @@ SET(LOGIC_SOURCES
 	wonko/OpSys.h
 	wonko/OpSys.cpp
 
+	# Resources
```
Comment (peterix) [inline logic/CMakeLists.txt]: I still feel this whole thing is like shooting mosquitos with an orbital death laser.

File: logic/CMakeLists.txt [hunk 17]
```diff
@@ -198,6 +234,12 @@ SET(LOGIC_SOURCES
 	tasks/ThreadTask.cpp
 	tasks/SequentialTask.h
 	tasks/SequentialTask.cpp
+	tasks/StandardTask.h
+	tasks/StandardTask.cpp
+	tasks/MoveFileTask.h
+	tasks/MoveFileTask.cpp
+	tasks/CopyFileTask.h
```
Comment (peterix) [inline logic/CMakeLists.txt]: These are actually never used and shouldn't be included.

File: logic/CMakeLists.txt [hunk 18]
```diff
@@ -249,3 +291,5 @@ target_link_libraries(MultiMC_logic iconfix libUtil LogicalGui ${QUAZIP_LIBRARIE
 		${MultiMC_LINK_ADDITIONAL_LIBS})
 
 add_dependencies(MultiMC_logic QuaZIP)
+
+include_directories(${CMAKE_CURRENT_SOURCE_DIR})
```
Comment (peterix) [inline logic/CMakeLists.txt]: The hell is this?

File: logic/MetaPackageList.cpp [hunk 19]
```diff
@@ -0,0 +1,315 @@
+/* Copyright 2013-2014 MultiMC Contributors
```
Comment (peterix) [inline logic/MetaPackageList.cpp]: Why is this file part of this pull request at all?

File: logic/auth/AccountModel.cpp [hunk 20]
```diff
@@ -0,0 +1,358 @@
...
+	using namespace Json;
+	const QJsonObject root = ensureObject(ensureDocument(data));
+
+	QList<BaseAccount *> accs;
+	QMap<QString, QMap<QString, BaseAccount *>> defs;
+
+	const int formatVersion = ensureInteger(root, "formatVersion", 0);
+	if (formatVersion == 2) // old, pre-multiauth format
+	{
+		const QString active = ensureString(root, "activeAccount", "");
```
Comment (peterix) [inline logic/auth/AccountModel.cpp]: This destroys the original account file.

Comment (peterix) [inline logic/auth/AccountModel.cpp]: I'm glad I have backups.

File: logic/auth/MojangAccount.cpp [hunk 21]
```diff
@@ -1,278 +0,0 @@
...
-	if (m_accessToken.isEmpty())
-		return NotVerified;
-	else
-		return Verified;
-}
-
-std::shared_ptr<YggdrasilTask> MojangAccount::login(AuthSessionPtr session,
-													QString password)
-{
-	Q_ASSERT(m_currentTask.get() == nullptr);
```
Comment (peterix) [inline logic/auth/MojangAccount.cpp]: I think you are entirely missing this logic, and you are changing how auth works, for the worse.

File: logic/auth/imgur/ImgurAccount.cpp [hunk 22]
```diff
@@ -0,0 +1,38 @@
+// Licensed under the Apache-2.0 license. See README.md for details.
```
Comment (peterix) [inline logic/auth/imgur/ImgurAccount.cpp]: Is this even used in any way?

File: logic/minecraft/JarMod.cpp [hunk 23]
```diff
@@ -0,0 +1 @@
+#include "JarMod.h"
```
Comment (peterix) [inline logic/minecraft/JarMod.cpp]: WHAT!?!

File: logic/minecraft/MinecraftPatch.cpp [hunk 24]
```diff
@@ -0,0 +1,219 @@
+#include <QJsonArray>
```
Comment (peterix) [inline logic/minecraft/MinecraftPatch.cpp]: This file being here makes absolutely no sense.

File: logic/minecraft/ProfileUtils.cpp [hunk 25]
```diff
@@ -9,6 +9,8 @@
 #include <QRegularExpression>
```
Comment (peterix) [inline logic/minecraft/ProfileUtils.cpp]: More nonsensical changes.

File: logic/minecraft/RawLibrary.cpp [hunk 26]
```diff
@@ -0,0 +1,93 @@
+#include "RawLibrary.h"
```
Comment (peterix) [inline logic/minecraft/RawLibrary.cpp]: And more nonsensical changes.

File: logic/minecraft/VersionBuildError.h [hunk 27]
```diff
@@ -3,7 +3,7 @@
 class VersionBuildError : public Exception
```
Comment (peterix) [inline logic/minecraft/VersionBuildError.h]: And more nonsensical changes.

File: logic/minecraft/auth/MojangAccount.cpp [hunk 28]
```diff
@@ -0,0 +1,224 @@
...
+	if (!hasToken("uuid") || token("uuid").isEmpty())
+	{
+		return QString();
+	}
+	return "web:https://crafatar.com/renders/body/" + token("uuid");
+}
+
+Task *MojangAccount::createLoginTask(const QString &username, const QString &password, SessionPtr session)
+{
+	if (accountStatus() == NotVerified && password.isEmpty())
```
Comment (peterix) [inline logic/minecraft/auth/MojangAccount.cpp]: This is missing the *INTELLIGENT* login method from the original.

File: logic/minecraft/auth/tasks/AuthenticateTask.cpp [hunk 29]
```diff
@@ -14,20 +14,15 @@
  * limitations under the License.
  */
 
-#include <auth/flows/AuthenticateTask.h>
```
Comment (peterix) [inline logic/minecraft/auth/tasks/AuthenticateTask.cpp]: I don't want to see any changes made to this. It was time tested and without a single known issue.

File: logic/minecraft/auth/tasks/RefreshTask.cpp [hunk 30]
```diff
@@ -13,18 +13,14 @@
  * limitations under the License.
  */
```
Comment (peterix) [inline logic/minecraft/auth/tasks/RefreshTask.cpp]: Same as AuthenticateTask.

File: logic/minecraft/auth/tasks/YggdrasilTask.cpp [hunk 31]
```diff
@@ -13,7 +13,7 @@
  * limitations under the License.
  */
```
Comment (peterix) [inline logic/minecraft/auth/tasks/YggdrasilTask.cpp]: Same as AuthenticateTask.

File: logic/minecraft/onesix/OneSixRead.cpp [hunk 32]
```diff
@@ -114,10 +115,11 @@ LibraryPtr readRawLibrary(const QJsonObject &libObj, const QString &filename)
 
 LibraryPtr OneSixFormat::readRawLibraryPlus(const QJsonObject &libObj, const QString &filename)
 {
+	using namespace Json;
 	LibraryPtr lib = readRawLibrary(libObj, filename);
 	if (libObj.contains("insert"))
 	{
-		QJsonValue insertVal = libObj.value("insert");
+		QJsonValue insertVal = ensureJsonValue(libObj.value("insert"), "library insert rule");
```
Comment (peterix) [inline logic/minecraft/onesix/OneSixRead.cpp]: Not acceptable.

File: logic/minecraft/onesix/OneSixWrite.cpp [hunk 33]
```diff
@@ -179,6 +181,6 @@ QJsonDocument OneSixFormat::toJson(PackagePtr file, bool saveOrder)
 QJsonObject OneSixFormat::toJson(JarmodPtr mod)
 {
 	QJsonObject out;
-	writeString(out, "name", mod->name);
```
Comment (peterix) [inline logic/minecraft/onesix/OneSixWrite.cpp]: Unrelated changes, not acceptable.

File: logic/minecraft/wonko/WonkoRead.cpp [hunk 34]
```diff
@@ -105,6 +103,8 @@ convertLibs(std::shared_ptr<Minecraft::Libraries> libs, const bool isNative)
 
 PackagePtr WonkoFormat::fromJson(const QJsonDocument &doc, const QString &filename)
 {
+	using namespace Json;
```
Comment (peterix) [inline logic/minecraft/wonko/WonkoRead.cpp]: Unrelated changes, not acceptable.

File: logic/tasks/CopyFileTask.h [hunk 35]
```diff
@@ -0,0 +1,25 @@
+// Licensed under the Apache-2.0 license. See README.md for details.
```
Comment (peterix) [inline logic/tasks/CopyFileTask.h]: Never used for anything substantial. Should be in a separate PR.

File: logic/tasks/MoveFileTask.cpp [hunk 36]
```diff
@@ -0,0 +1,33 @@
+// Licensed under the Apache-2.0 license. See README.md for details.
```
Comment (peterix) [inline logic/tasks/MoveFileTask.cpp]: Same as CopyFileTask.

File: logic/tasks/ProgressProvider.h [hunk 37]
```diff
@@ -23,6 +23,8 @@ class ProgressProvider : public QObject
 protected:
 	explicit ProgressProvider(QObject *parent = 0) : QObject(parent)
 	{
+		connect(this, &ProgressProvider::succeeded, this, &ProgressProvider::finished);
```
Comment (peterix) [inline logic/tasks/ProgressProvider.h]: I assume there is some reason for this?

File: logic/tasks/StandardTask.cpp [hunk 38]
```diff
@@ -0,0 +1,130 @@
+// Licensed under the Apache-2.0 license. See README.md for details.
```
Comment (peterix) [inline logic/tasks/StandardTask.cpp]: This file should be turned into a task factory for standard tasks, or maybe a bunch of static Something::create methods.

File: logic/tasks/StandardTask.cpp [hunk 39]
```diff
@@ -0,0 +1,130 @@
...
+	if (other->isRunning())
+	{
+		m_loop->exec();
+	}
+	disconnect(other.get(), 0, m_loop, 0);
+	disconnect(other.get(), 0, this, 0);
+	other->deleteLater();
+	if (!other->successful())
+	{
+		throw Exception(other->failReason());
```
Comment (peterix) [inline logic/tasks/StandardTask.cpp]: No.

File: logic/tasks/Task.h [hunk 40]
```diff
@@ -40,6 +40,8 @@ class Task : public ProgressProvider
 	 */
 	virtual QString failReason() const;
 
+	virtual bool  canAbort() const { return false; }
```
Comment (peterix) [inline logic/tasks/Task.h]: Interesting. That would be a step in a right, although unrelated direction.

File: tests/ModelTester.h [hunk 41]
```diff
@@ -0,0 +1,72 @@
+#pragma once
+
+#include <QAbstractItemModel>
+#include <QTest>
+#include <memory>
+
+// does some basic sanity checking (doesn't crash on invalid indexes etc.)
+class ModelTester : public QObject
```
Comment (peterix) [inline tests/ModelTester.h]: Not sure how this is testing anything our small city's worth of users wouldn't discover in 5 seconds, but OK ;)

**Comment by peterix** [comment]
Alright, something that's hard to pin to any one line of code:
* ate my original account file. Had to restore from backup.
* I found the account file saved to my build folder. There's something very wrong with it ending up in odd places.
* Confuses user login with user name, does not save actual user login. This is VERY annoying to deal with for users. I would be annoyed by having to type in my e-mail in front of a stream.
* There is no way to cancel login. Any error is taken as 'HEY USER. LOGIN OR I'LL KEEP YOU HERE FOREVER'. Solved by `kill -9`. You rewrote the perfectly good logic for logins in MainWindow and replaced it with something that doesn't work.
* There seems to be no way to reset the default account to none. It's gone from the main window.
* There's no way to reliably tell which account is default in the accounts page. It's just all the same. The button/checkbox mess doesn't help.
* Some other features seem to have hone missing. Like actually using the right yggdrasil task to login? Not going to go and do a thorough analysis beyond that. There's too much wrong already.
* Account page has extra ugly margins (should be 0).
* Account page view needs work - default left column size obstructs user names.
* Maybe the avatar thingy could face the rest of the window and not have visible seams? Just personal preference.
* Overhead: adds many classes only used for auth.


I think more account types are a good idea. 
I also think the good parts of this can be added with a lot less code.

All in all, very raw.

**Comment by peterix** [comment]
Also, which part of this allows setting accounts per instance? I wasn't able to find that at all.

**Comment by peterix** [comment]
Needs rework.

File: logic/CMakeLists.txt (hunk 16 above)
Comment (peterix) [inline logic/CMakeLists.txt]: Probably should be introduced slower, and in a different way.

File: application/AuthTask.cpp (hunk 1 above)
Comment (02JanDal) [inline application/AuthTask.cpp]: This task was originally in logic, and used LogicalGui to show stuff. You told me to not use LogicalGui, which required me to put GUI in logic and the entire thing into application/.

File: .travis.yml (hunk 2 above)
Comment (02JanDal) [inline .travis.yml]: Not sure, will have to check.

File: application/AuthTask.cpp (hunk 3 above)
Comment (02JanDal) [inline application/AuthTask.cpp]: ???

How is this any different from parsing JSON in a task and catching the errors from that?

File: application/AuthTask.cpp (hunk 4 above)
Comment (02JanDal) [inline application/AuthTask.cpp]: Protip: Check the code before commenting. (in this case, check StandardTask::runTask)

File: application/CMakeLists.txt (hunk 5 above)
Comment (02JanDal) [inline application/CMakeLists.txt]: Not entirely sure, will have to check.

File: application/MainWindow.cpp (hunk 6 above)
Comment (02JanDal) [inline application/MainWindow.cpp]: The `while(true)` loop is bad indeed, but something I had to do to shoehorn the MojangAuthSession thingy into the picture. Not being able to exit is bad too. I'll see if I can clean this up.

File: application/MainWindow.cpp (hunk 7 above)
Comment (02JanDal) [inline application/MainWindow.cpp]: I think Github is derping something here.

```cpp
case MojangAuthSession::RequiresPassword:
{
	qCritical() << "This should not be reachable";
	break;
}
```

If a password is needed the task will fail, meaning ProgressDialog::exec returns QDialog::Rejected which returns on MainWindow.cpp#1512

File: application/MainWindow.cpp (hunk 8 above)
Comment (02JanDal) [inline application/MainWindow.cpp]: Weird, probably some artifact of the merging.

File: application/dialogs/AccountLoginDialog.cpp (hunk 9 above)
Comment (02JanDal) [inline application/dialogs/AccountLoginDialog.cpp]: Yes? This calls reject() when the USER clicks cancel?

File: application/dialogs/AccountSelectDialog.cpp (hunk 10 above)
Comment (02JanDal) [inline application/dialogs/AccountSelectDialog.cpp]: It was replaced by AccountsDialog. The accounts page and AccountSelectDialog where in functionality and look almost identical. Now there is instead an AccountsWidget that is used by both the dialog and the page.

File: application/pages/VersionPage.cpp (hunk 11 above)
Comment (02JanDal) [inline application/pages/VersionPage.cpp]: Merge artifact I'd say (I quite hope you realize that yourself)

File: application/widgets/AccountsWidget.ui (hunk 12 above)
Comment (02JanDal) [inline application/widgets/AccountsWidget.ui]: Will fix

File: application/widgets/AccountsWidget.ui (hunk 13 above)
Comment (02JanDal) [inline application/widgets/AccountsWidget.ui]: Will check

File: logic/AbstractCommonModel.cpp (hunk 14 above)
Comment (02JanDal) [inline logic/AbstractCommonModel.cpp]: Not having to implement a QAbstractItemModel everywhere. It is currently only used for AccountTypesModel (AccountModel.cpp#18-27), but if you take a look at it and compare to what you'd need to do without it you'll see the use of it. This is roughly what would be needed otherwise:

```cpp
class AccountTypesModel : public QAbstractListModel
{
    Q_OBJECT
public:
    explicit AccountTypesModel(QObject *parent = nullptr) : QAbstractItemModel(parent) {}

    int rowCount() const override
    {
        return m_rows.size();
    }
    QVariant data(const QModelIndex &index, const int role) const override
    {
        BaseAccountType *row = m_rows.at(index.row());
        switch (role)
        {
        case Qt::DisplayRole:
            return row->text();
        case Qt::UserRole:
            return row->id();
        case Qt::DecorationRole:
            return row->icon();
        }
        return QVariant();
    }

private:
    QList<BaseAccountType *> m_rows;
};
```
AccountTypesModel is even relatively simple. Add headers, multiple columns etc. and you have way more code. Also should be noted that the above does not have insertion/deletion, adding that would further increase the boilerplate code.

Having this will make more sense when used in more places, but I wanted to keep the changes low.

File: logic/BaseInstance.h (hunk 15 above)
Comment (02JanDal) [inline logic/BaseInstance.h]: Yay merge artifacts (I really hope this can get into one of the main branches soon :P)

File: logic/CMakeLists.txt (hunk 16 above)
Comment (02JanDal) [inline logic/CMakeLists.txt]: Again, I could go ahead and use this for other things, but that would mean a bigger diff...

File: logic/CMakeLists.txt (hunk 17 above)
Comment (02JanDal) [inline logic/CMakeLists.txt]: True for Copy and Move, will remove.

File: logic/CMakeLists.txt (hunk 18 above)
Comment (02JanDal) [inline logic/CMakeLists.txt]: Again, not sure.

File: logic/MetaPackageList.cpp (hunk 19 above)
Comment (02JanDal) [inline logic/MetaPackageList.cpp]: I'll let you guess

File: logic/auth/AccountModel.cpp (hunk 20 above)
Comment (02JanDal) [inline logic/auth/AccountModel.cpp]: WTF? I'm 95% sure I had a file copy in there

File: logic/auth/MojangAccount.cpp (hunk 21 above)
Comment (02JanDal) [inline logic/auth/MojangAccount.cpp]: I'm not entirely sure what you mean here since Github seems to have missed that this file moved...

File: logic/auth/imgur/ImgurAccount.cpp (hunk 22 above)
Comment (02JanDal) [inline logic/auth/imgur/ImgurAccount.cpp]: Nope, miss-clicked

File: logic/minecraft/JarMod.cpp (hunk 23 above)
Comment (02JanDal) [inline logic/minecraft/JarMod.cpp]: Merge artifact

File: logic/minecraft/MinecraftPatch.cpp (hunk 24 above)
Comment (02JanDal) [inline logic/minecraft/MinecraftPatch.cpp]: Merge artifact

File: logic/minecraft/ProfileUtils.cpp (hunk 25 above)
Comment (02JanDal) [inline logic/minecraft/ProfileUtils.cpp]: Merge artifact

File: logic/minecraft/RawLibrary.cpp (hunk 26 above)
Comment (02JanDal) [inline logic/minecraft/RawLibrary.cpp]: Merge artifact

File: logic/minecraft/VersionBuildError.h (hunk 27 above)
Comment (02JanDal) [inline logic/minecraft/VersionBuildError.h]: Merge artifact

File: logic/minecraft/auth/MojangAccount.cpp (hunk 28 above)
Comment (02JanDal) [inline logic/minecraft/auth/MojangAccount.cpp]: What was intelligent about it? I just double checked and there was nothing that told me about what it did special (*cough* comments? *cough*). I know I'm not good at adding comments either, but if there is an non-obvious method doing something fancy it should be marked.

File: logic/minecraft/auth/tasks/AuthenticateTask.cpp (hunk 29 above)
Comment (02JanDal) [inline logic/minecraft/auth/tasks/AuthenticateTask.cpp]: The changes are either for adapting to changes in code elsewhere (amongst others not writing directly to member variables, yuck) or cleaning it up (adding braces around if, using ensureX). The changes will stay. According to your logic we shouldn't be doing wonko either.

File: logic/minecraft/auth/tasks/RefreshTask.cpp (hunk 30 above)
Comment (02JanDal) [inline logic/minecraft/auth/tasks/RefreshTask.cpp]: Same as AuthenticateTask

File: logic/minecraft/auth/tasks/YggdrasilTask.cpp (hunk 31 above)
Comment (02JanDal) [inline logic/minecraft/auth/tasks/YggdrasilTask.cpp]: Same as AuthenticateTask

File: logic/minecraft/onesix/OneSixRead.cpp (hunk 32 above)
Comment (02JanDal) [inline logic/minecraft/onesix/OneSixRead.cpp]: I see your point, though I only half agree with it. I'll decide if I'll change this later.

File: application/MainWindow.cpp (hunk 7 above)
Comment (peterix) [inline application/MainWindow.cpp]: Except it doesn't.

File: application/dialogs/AccountLoginDialog.cpp (hunk 9 above)
Comment (peterix) [inline application/dialogs/AccountLoginDialog.cpp]: And then it doesn't work.

File: logic/minecraft/onesix/OneSixWrite.cpp (hunk 33 above)
Comment (02JanDal) [inline logic/minecraft/onesix/OneSixWrite.cpp]: Guess what happens if I revert this? (it's possible that it was fixed after Json.{h,cpp} got included in unstable, I'll check later)

File: logic/minecraft/wonko/WonkoRead.cpp (hunk 34 above)
Comment (02JanDal) [inline logic/minecraft/wonko/WonkoRead.cpp]: Same as above

File: logic/tasks/CopyFileTask.h (hunk 35 above)
Comment (02JanDal) [inline logic/tasks/CopyFileTask.h]: As mentioned somewhere above, agreed

File: logic/tasks/MoveFileTask.cpp (hunk 36 above)
Comment (02JanDal) [inline logic/tasks/MoveFileTask.cpp]: As mentioned somewhere above, agreed

File: logic/tasks/ProgressProvider.h (hunk 37 above)
Comment (02JanDal) [inline logic/tasks/ProgressProvider.h]: There is indeed. There are cases when you don't care how something finished, you just care about it being done.

File: logic/tasks/StandardTask.cpp (hunk 38 above)
Comment (02JanDal) [inline logic/tasks/StandardTask.cpp]: There are two interesting parts in this one, neither related to factories really:
1. runTask and runTaskNonBlocking (should probably be called runTaskAsync or something). They will stay here whatever happens.
2. The others. Some are just wrappers around another task, in which case they can be removed. Others do a bit more (networkGetCached etc.), some more thought will need to be put into them.

File: logic/tasks/StandardTask.cpp (hunk 39 above)
Comment (02JanDal) [inline logic/tasks/StandardTask.cpp]: "properly"...?

File: tests/ModelTester.h (hunk 41 above)
Comment (02JanDal) [inline tests/ModelTester.h]: It's mainly for development and derps. You usually won't run into any of these issues in the wild.

**Comment by 02JanDal** [comment]
* ate my original account file. Had to restore from backup.
    * as noted in inline comments, will fix
* I found the account file saved to my build folder. There's something very wrong with it ending up in odd places.
    * It gets saved to whereever QSaveFile asdf("accounts.json") decides to save it, which should be relative to QDir::current()
* Confuses user login with user name, does not save actual user login. This is VERY annoying to deal with for users. I would be annoyed by having to type in my e-mail in front of a stream.
    * It saves it, but as pointed out on IRC yesterday something about that broke. Will fix
    * I'm only 50% about hiding it. I can see it being useful though, so I guess I'll do that
* There is no way to cancel login. Any error is taken as 'HEY USER. LOGIN OR I'LL KEEP YOU HERE FOREVER'. Solved by `kill -9`. You rewrote the perfectly good logic for logins in MainWindow and replaced it with something that doesn't work.
    * I'll have to check what's causing this
    * "good logic ... in MainWindow" <- something's wrong about that sentence...
* There seems to be no way to reset the default account to none. It's gone from the main window.
    * Unchecking the checkbox should do it, I did change something related to that recently though, will check.
    * Option in mainwindow: yep, it's gone. I could add it back, but I'm not really sure how useful it is.
* There's no way to reliably tell which account is default in the accounts page. It's just all the same. The button/checkbox mess doesn't help.
    * As discussed on IRC: extra column. And as said before: I suck at UI.
* Some other features seem to have hone missing. Like actually using the right yggdrasil task to login? Not going to go and do a thorough analysis beyond that. There's too much wrong already.
    * I'll do a check through and see what I can find
* Account page has extra ugly margins (should be 0).
    * See inline comment
* Account page view needs work - default left column size obstructs user names.
    * Will fix
* Maybe the avatar thingy could face the rest of the window and not have visible seams? Just personal preference.
    * I agree fully, it would need either a change in crafatar or adding some mirroring logic.
* Overhead: adds many classes only used for auth.
    * Ehh... yes? There are a few base classes? What did you expect?

Setting defaults per instance is I think only possible when launching an instance without having a default set. Yes this is bad and I'll fix that.

File: application/MainWindow.cpp (hunk 7 above)
Comment (02JanDal) [inline application/MainWindow.cpp]: Well, it _should_. I'll check.

File: application/dialogs/AccountLoginDialog.cpp (hunk 9 above)
Comment (02JanDal) [inline application/dialogs/AccountLoginDialog.cpp]: It _should_ work. I'll check.

File: application/MainWindow.cpp (hunk 8 above)
Comment (02JanDal) [inline application/MainWindow.cpp]: Wasn't a merge issue, see https://github.com/MultiMC/MultiMC5/pull/889/files#diff-1b1b7cec09f1821effcfc681e3226cd5R1593

**Comment by 02JanDal** [comment]
Some top level things that need fixing:
* BUGS BUGS BUGS BUGS (yes, there are bugs)
    * ...and barely tested with multiple accounts
    * It requires a password after every restart???
* UI design. I'm terrible at it.

@peterix are there any more major flaws? (so that I can fix them before I fix bugs)

**Comment by peterix** [comment]
Rebase on top of develop should be done first.

**Comment by 02JanDal** [comment]
Closing because of the changed target branch.
